Expand app quick start guides

azigler opened the pull request:
**Fixes:**
- https://mattermost.atlassian.net/browse/MM-39146
- https://mattermost.atlassian.net/browse/MM-36715
- https://mattermost.atlassian.net/browse/MM-37115
- https://github.com/mattermost/mattermost-developer-documentation/issues/942
- https://github.com/mattermost/mattermost-developer-documentation/issues/1028
- https://github.com/mattermost/mattermost-developer-documentation/issues/937

**Adds:**
- Developer setup guide using new https://github.com/mattermost/mattermost-app-examples repository
- New Python quick start guide for the app framework based on https://github.com/mattermost/mattermost-plugin-apps/pull/352
- Refreshed TypeScript Hello World guide
- Refreshed Go Hello World guide
- Refreshed Go Serverless guide
- New Golang JWT guide

**Blocked guides:**
- Golang
  - Lifecycle - Go error when trying to install and use lifecycle bindings.
  - Oauth2 - Requires the entire dev setup to be exposed to the web via something like `ngrok`, will require more thought.
  - Webhooks - `/hello-webhooks info` endpoint does nothing when sent to, app does not report messages to channel.

azigler commented:
/update-branch

azigler commented:
/update-branch

azigler commented:
Easy links to view the new pages:
- [Developer setup](http://mattermost-dev-docs-preview-pulls.s3-website-us-east-1.amazonaws.com/1106/integrate/apps/quickstart/)
- [Hello World in Go](http://mattermost-dev-docs-preview-pulls.s3-website-us-east-1.amazonaws.com/1106/integrate/apps/quickstart/quick-start-go/)
- [Hello World in Python](http://mattermost-dev-docs-preview-pulls.s3-website-us-east-1.amazonaws.com/1106/integrate/apps/quickstart/quick-start-python/)
- [JWT in Go](http://mattermost-dev-docs-preview-pulls.s3-website-us-east-1.amazonaws.com/1106/integrate/apps/quickstart/quick-start-jwt/)
- [Serverless on AWS or OpenFaaS in Go](http://mattermost-dev-docs-preview-pulls.s3-website-us-east-1.amazonaws.com/1106/integrate/apps/quickstart/quick-start-serverless/)
- [Hello World in Go](http://mattermost-dev-docs-preview-pulls.s3-website-us-east-1.amazonaws.com/1106/integrate/apps/quickstart/quick-start-js/)

And if you haven't checked it out yet, here's the shiny new repo that houses all these examples, their `docker-compose.yml`s, and the new dev server `docker-compose.yml` that makes it super easy to spin up the demos: https://github.com/mattermost/mattermost-app-examples

@levb @mickmister I'd appreciate your technical review on these guides, to make sure I didn't botch anything. You'll see I added a new guide for Python, too.

@mickmister Just a heads up that I noticed some issues with the **lifecycle** example (I got an error when trying to install and use lifecycle bindings with that example) and the **webhooks** example (`/hello-webhooks` info endpoint does nothing when sent to, app does not report messages to channel). If you can help me figure out/unblock those issues and we can get them working again, I can add guides for them in a future PR. The **OAuth2** example requires the entire dev setup to be exposed to the web via something like ngrok, so it will require more thought for a future guide.

@cwarnermm @BenLloydPearson I tagged y'all as reviewers on this, if you don't mind taking a look based on the many conversations we've had about these guides. Carrie, can you stick the appropriate review tags on this PR? 🙂 Also, I noticed the TypeScript guide's URL is "js" which isn't a big deal, but any chance we could change it to "ts" and have "js" forward to it? It would be nice for the URL to match the language. Also, if you can figure out how to weight the guides so the three Hello Worlds are first, followed by serverless, followed by JWT, that would be *amazing!* Then we can re-order the index table on the developer guide, too.

neflyte commented on site/content/integrate/apps/quickstart/quick-start-python/_index.md:
```diff
@@ -0,0 +1,93 @@
+  - /integrate/apps/quick-start-python/
+---
+
+This quick start guide will walk you through the basics of creating a hello world Mattermost app in Python. In this guide you will review an app that:
```
```suggestion
This quick start guide will walk you through the basics of creating a hello world Mattermost App in Python. In this guide you will review an App that:
```

neflyte commented on site/content/integrate/apps/quickstart/quick-start-python/_index.md:
```diff
@@ -0,0 +1,93 @@
+
+## Prerequisites
+
+Before you can start with your app, you first need to set up your environment by following the [developer setup guide]({{< ref "/integrate/apps/quickstart" >}}).
```
```suggestion
Before you can start with your App, you first need to set up your environment by following the [developer setup guide]({{< ref "/integrate/apps/quickstart" >}}).
```

neflyte commented on site/content/integrate/apps/quickstart/quick-start-python/_index.md:
```diff
@@ -0,0 +1,93 @@
+
+You also need [Python version 3 installed](https://www.python.org/downloads/).
+
+## Download and start the app
```
```suggestion
## Download and start the App
```

neflyte commented on site/content/integrate/apps/quickstart/quick-start-python/_index.md:
```diff
@@ -0,0 +1,93 @@
+
+## Download and start the app
+
+In the same [mattermost-app-examples repository](https://github.com/mattermost/mattermost-app-examples) you cloned via the [developer setup guide]({{< ref "/integrate/apps/quickstart" >}}) above, navigate to the [`python/hello-world`](https://github.com/mattermost/mattermost-app-examples/tree/master/python/hello-world) directory and start the Docker compose:
```
```suggestion
In the same [mattermost-app-examples repository](https://github.com/mattermost/mattermost-app-examples) you cloned via the [developer setup guide]({{< ref "/integrate/apps/quickstart" >}}) above, navigate to the [`python/hello-world`](https://github.com/mattermost/mattermost-app-examples/tree/master/python/hello-world) directory and start the Docker Compose stack:
```

neflyte commented on site/content/integrate/apps/quickstart/quick-start-python/_index.md:
````diff
@@ -0,0 +1,93 @@
+
+```sh
+cd python/hello-world
+docker compose up
````
```suggestion
docker-compose up
```

neflyte commented on site/content/integrate/apps/quickstart/quick-start-python/_index.md:
````diff
@@ -0,0 +1,93 @@
+docker compose up
+```
+
+You'll see Docker install the npm modules and then the app will come online and print the following message:
````
Python doesn't use `npm`.

azigler replied:
Great catch, thanks!

neflyte commented on site/content/integrate/apps/quickstart/quick-start-python/_index.md:
````diff
@@ -0,0 +1,93 @@
+Running on http://mattermost-apps-python-hello-world:8090
+```
+
+## Install the app on Mattermost
````
```suggestion
## Install the App on Mattermost
```

neflyte commented on site/content/integrate/apps/quickstart/quick-start-python/_index.md:
```diff
@@ -0,0 +1,93 @@
+
+## Install the app on Mattermost
+
+Next, access your development Mattermost server at [http://localhost:8065](http://localhost:8065) and use the `/apps install http http://mattermost-apps-python-hello-world:8090/manifest.json` slash command to install the hello world app. Select `Agree to grant the app access to APIs and Locations` and click `Submit` to finish the installation.
```
```suggestion
Next, access your development Mattermost Server at [http://localhost:8065](http://localhost:8065) and use the `/apps install http http://mattermost-apps-python-hello-world:8090/manifest.json` slash command to install the hello world App. Select `Agree to grant the app access to APIs and Locations` and click `Submit` to finish the installation.
```

neflyte commented on site/content/integrate/apps/quickstart/quick-start-python/_index.md:
```diff
@@ -0,0 +1,93 @@
+
+Next, access your development Mattermost server at [http://localhost:8065](http://localhost:8065) and use the `/apps install http http://mattermost-apps-python-hello-world:8090/manifest.json` slash command to install the hello world app. Select `Agree to grant the app access to APIs and Locations` and click `Submit` to finish the installation.
+
+## Use the app
```
```suggestion
## Use the App
```

neflyte commented on site/content/integrate/apps/quickstart/quick-start-python/_index.md:
```diff
@@ -0,0 +1,93 @@
+
+![image](first.png)
+
+## Review the app
```
```suggestion
## Review the App
```

neflyte commented on site/content/integrate/apps/quickstart/quick-start-python/_index.md:
```diff
@@ -0,0 +1,93 @@
+
+## Review the app
+
+To understand the app, examine the following elements:
```
```suggestion
To understand the App, examine the following elements:
```

neflyte commented on site/content/integrate/apps/quickstart/quick-start-python/_index.md:
```diff
@@ -0,0 +1,93 @@
+
+### Manifest
+
+The app has to provide a manifest, which declares the app's metadata required for installation. In this example, the following permissions and locations are requested:
```
```suggestion
The App has to provide a manifest, which declares the App's metadata required for installation. In this example, the following permissions and locations are requested:
```

neflyte commented on site/content/integrate/apps/quickstart/quick-start-python/_index.md:
```diff
@@ -0,0 +1,93 @@
+
+### Bindings and locations
+
+This app adds a channel header button and slash commands. In order to register these locations, there is a `POST` handler for the `/bindings` endpoint on your app's API.
```
```suggestion
This App adds a channel header button and slash commands. In order to register these locations, there is a `POST` handler for the `/bindings` endpoint in your App's API.
```

neflyte commented on site/content/integrate/apps/quickstart/quick-start-python/_index.md:
```diff
@@ -0,0 +1,93 @@
+
+### Functions and forms
+
+Forms handle user events on the bindings. This app provides a `form` before `POST`ing to the `/submit` function. In the case of a channel header, the form will launch a modal to collect its fields. In the case of a slash command, the form's fields will be collected as arguments from the user's command.
```
```suggestion
Forms handle user events on the bindings. This App provides a `form` before `POST`ing to the `/submit` function. In the case of a channel header, the form will launch a modal to collect its fields. In the case of a slash command, the form's fields will be collected as arguments from the user's command.
```

neflyte commented on site/content/integrate/apps/quickstart/quick-start-python/_index.md:
```diff
@@ -0,0 +1,93 @@
+
+### Assets
+
+Apps may include static assets (e.g., `icon.png`). Static assets must be served under the `static` path. For example, there is a `icon.png` file in the `src/static` directory that this app serves.
```
```suggestion
Apps may include static assets (e.g., `icon.png`). Static assets must be served under the `static` path. For example, there is a `icon.png` file in the `src/static` directory that this App serves.
```

neflyte commented on site/content/integrate/apps/quickstart/quick-start-python/_index.md:
```diff
@@ -0,0 +1,93 @@
+
+Apps may include static assets (e.g., `icon.png`). Static assets must be served under the `static` path. For example, there is a `icon.png` file in the `src/static` directory that this app serves.
+
+## Uninstall the app
```
```suggestion
## Uninstall the App
```

neflyte commented on site/content/integrate/apps/quickstart/quick-start-python/_index.md:
```diff
@@ -0,0 +1,93 @@
+
+## Uninstall the app
+
+Once you're done with the app, you can uninstall it via the `/apps uninstall hello-world` slash command. Alternatively, you can use `/apps debug clean` to remove all data for all installed apps.
```
```suggestion
Once you're done with the App, you can uninstall it via the `/apps uninstall hello-world` slash command. Alternatively, you can use `/apps debug clean` to remove all data for all installed Apps.
```

neflyte commented on site/content/integrate/apps/quickstart/quick-start-python/_index.md:
```diff
@@ -0,0 +1,93 @@
+
+Once you're done with the app, you can uninstall it via the `/apps uninstall hello-world` slash command. Alternatively, you can use `/apps debug clean` to remove all data for all installed apps.
+
+To stop and clean up the app from Docker after you're done, use the following command in the `python/hello-world` directory:
```
```suggestion
To stop and clean up the App from Docker after you're done, use the following command in the `python/hello-world` directory:
```

neflyte commented on site/content/integrate/apps/quickstart/quick-start-python/_index.md:
````diff
@@ -0,0 +1,93 @@
+To stop and clean up the app from Docker after you're done, use the following command in the `python/hello-world` directory:
+
+```sh
+docker compose down
````
```suggestion
docker-compose down
```

neflyte commented on site/content/integrate/apps/quickstart/quick-start-python/_index.md:
```diff
@@ -0,0 +1,93 @@
+
+## Conclusion
+
+You now know how to create a Mattermost app in Python. If you have questions about building apps or want to show off what you're building, join us on the [Integrations & Apps channel in the Mattermost Community server](https://community.mattermost.com/core/channels/integrations)!
```
```suggestion
You now know how to create a Mattermost App in Python. If you have questions about building Apps or want to show off what you're building, join us on the [Integrations & Apps channel in the Mattermost Community server](https://community.mattermost.com/core/channels/integrations)!
```

neflyte commented on site/content/integrate/apps/quickstart/quick-start-serverless/_index.md:
```diff
@@ -1,436 +1,93 @@
-description: "This quick start guide will walk you through the basics of writing a Mattermost App that can be deployed to serverless platforms, in Go."
-weight: 30
+title: "Serverless on AWS or OpenFaaS in Go"
+heading: "Create a serverless Mattermost app on AWS or OpenFaaS in Go"
```
```suggestion
heading: "Create a serverless Mattermost App on AWS or OpenFaaS in Go"
```

neflyte commented on site/content/integrate/apps/quickstart/quick-start-serverless/_index.md:
```diff
@@ -1,436 +1,93 @@
-weight: 30
+title: "Serverless on AWS or OpenFaaS in Go"
+heading: "Create a serverless Mattermost app on AWS or OpenFaaS in Go"
+description: "This quick start guide will walk you through the basics of using creating a serverless Mattermost app in Go on AWS or OpenFaaS."
```
```suggestion
description: "This quick start guide will walk you through the basics of using creating a serverless Mattermost App in Go on AWS or OpenFaaS."
```

neflyte commented on site/content/integrate/apps/quickstart/quick-start-serverless/_index.md:
```diff
@@ -1,436 +1,93 @@
-- [`Makefile`](#7-makefile) that can build aws and openfaas bundles (`make dist`).
-
-See complete source code of the example [here](https://github.com/mattermost/mattermost-plugin-apps/tree/master/examples/go/hello-serverless).
+This quick start guide will walk you through the basics of using creating a serverless Mattermost app in Go on AWS or OpenFaaS. In this guide you will review an app that:
```
```suggestion
This quick start guide will walk you through the basics of using creating a serverless Mattermost App in Go on AWS or OpenFaaS. In this guide you will review an App that:
```

neflyte commented on site/content/integrate/apps/quickstart/quick-start-serverless/_index.md:
```diff
@@ -1,436 +1,93 @@
+- Attaches `send` function to a `/hello-serverless` command, accepting an optional `message` argument.
 
-Before you can start with your app, you first need to set up a local developer environment following the [server]({{< ref "/contribute/server/developer-setup" >}}) and [webapp]({{< ref "/contribute/webapp/developer-setup" >}}) setup guides. You must enable the apps feature flag before starting the Mattermost server by setting the environment variable `MM_FEATUREFLAGS_AppsEnabled` to `true` by e.g ...
+This app will focus on the functionality of the actual serverless app in this example. For more information about [how to package a Mattermost app for serverless deployments]({{< ref "/integrate/apps/deploy/package-for-aws" >}}), [how to deploy a Mattermost app to AWS]({{< ref "/integrate/apps/deploy/deploy-aws" >}}), or [how to deploy a Mattermost app to OpenFaaS]({{< ref "/integrate/apps/deploy ...
```
```suggestion
This App will focus on the functionality of the actual serverless App in this example. For more information about [how to package a Mattermost App for serverless deployments]({{< ref "/integrate/apps/deploy/package-for-aws" >}}), [how to deploy a Mattermost App to AWS]({{< ref "/integrate/apps/deploy/deploy-aws" >}}), or [how to deploy a Mattermost App to OpenFaaS]({{< ref "/integrate/apps/deploy/deploy-openfaas" >}}) please refer to the corresponding guides.
```

neflyte commented on site/content/integrate/apps/quickstart/quick-start-serverless/_index.md:
```diff
@@ -1,436 +1,93 @@
 
-- `Enable Bot Account Creation`
-- `Enable OAuth 2.0 Service Provider`
+Before you can start with your app, you first need to set up your environment by following the [developer setup guide]({{< ref "/integrate/apps/quickstart" >}}).
```
```suggestion
Before you can start with your App, you first need to set up your environment by following the [developer setup guide]({{< ref "/integrate/apps/quickstart" >}}).
```

neflyte commented on site/content/integrate/apps/quickstart/quick-start-serverless/_index.md:
```diff
@@ -1,436 +1,93 @@
+Before you can start with your app, you first need to set up your environment by following the [developer setup guide]({{< ref "/integrate/apps/quickstart" >}}).
 
-You also need at least `go1.16` installed. Please follow the guide [here](https://golang.org/doc/install) to install the latest version.
+You also need at least `go1.16` installed. Please follow the [official guide](https://golang.org/doc/install) to install the latest version.
```
```suggestion
You also need Go v1.16 or later installed. Please follow the [official guide](https://golang.org/doc/install) to install the latest version.
```

neflyte commented on site/content/integrate/apps/quickstart/quick-start-serverless/_index.md:
```diff
@@ -1,436 +1,93 @@
+You also need at least `go1.16` installed. Please follow the [official guide](https://golang.org/doc/install) to install the latest version.
 
-### Install the Apps plugin
+## Download and start the app
```
```suggestion
## Download and start the App
```

neflyte commented on site/content/integrate/apps/quickstart/quick-start-serverless/_index.md:
```diff
@@ -1,436 +1,93 @@
+## Download and start the app
 
-The [apps plugin](https://github.com/mattermost/mattermost-plugin-apps) is a communication bridge between your app and the Mattermost server. To install it on your local server, start by cloning the code in a directory of your choice run:
+In the same [mattermost-app-examples repository](https://github.com/mattermost/mattermost-app-examples) you cloned via the [developer setup guide]({{< ref "/integrate/apps/quickstart" >}}) above, navigate to the [`golang/serverless`](https://github.com/mattermost/mattermost-app-examples/tree/master/golang/serverless) directory and start the Docker compose:
```
```suggestion
In the same [mattermost-app-examples repository](https://github.com/mattermost/mattermost-app-examples) you cloned via the [developer setup guide]({{< ref "/integrate/apps/quickstart" >}}) above, navigate to the [`golang/serverless`](https://github.com/mattermost/mattermost-app-examples/tree/master/golang/serverless) directory and start the Docker Compose stack:
```

neflyte commented on site/content/integrate/apps/quickstart/quick-start-serverless/_index.md:
````diff
@@ -1,436 +1,93 @@
-git clone https://github.com/mattermost/mattermost-plugin-apps.git
+```sh
+cd golang/serverless
+docker compose up
````
```suggestion
docker-compose up
```

neflyte commented on site/content/integrate/apps/quickstart/quick-start-serverless/_index.md:
```diff
@@ -1,436 +1,93 @@
-## Build the App
-
-Start building your app by creating a directory for the code and setting up a new go module:
+The `docker-compose.yml` uses the [HTTP deployment]({{< ref "/integrate/apps/deploy/deploy-http/" >}}) of the serverless app for the sake of trying it out locally. You'll see Docker install the Go modules and then the app will come online and print the following message:
```
```suggestion
The `docker-compose.yml` uses the [HTTP deployment]({{< ref "/integrate/apps/deploy/deploy-http/" >}}) of the serverless App for the sake of trying it out locally. You'll see Docker install the Go modules and then the App will come online and print the following message:
```

neflyte commented on site/content/integrate/apps/quickstart/quick-start-serverless/_index.md:
````diff
@@ -1,436 +1,93 @@
-	http.DefaultServeMux.ServeHTTP(w, r)
-}
-```
+## Install the app on Mattermost
````
```suggestion
## Install the App on Mattermost
```

neflyte commented on site/content/integrate/apps/quickstart/quick-start-serverless/_index.md:
```diff
@@ -1,436 +1,93 @@
 
-Note: names `package function` and `func Handle` appear to be required
-(hardcoded) for OpenFaaS.
+Next, access your development Mattermost server at [http://localhost:8065](http://localhost:8065) and use the `/apps install http http://mattermost-apps-golang-serverless:8080/manifest.json ` slash command to install the webhooks app. Select `Agree to grant the app access to APIs and Locations` and click `Submit` to finish the installation.
```
```suggestion
Next, access your development Mattermost Server at [http://localhost:8065](http://localhost:8065) and use the `/apps install http http://mattermost-apps-golang-serverless:8080/manifest.json ` slash command to install the webhooks App. Select `Agree to grant the app access to APIs and Locations` and click `Submit` to finish the installation.
```

neflyte commented on site/content/integrate/apps/quickstart/quick-start-serverless/_index.md:
```diff
@@ -1,436 +1,93 @@
 
-Note: names `package function` and `func Handle` appear to be required
-(hardcoded) for OpenFaaS.
+Next, access your development Mattermost server at [http://localhost:8065](http://localhost:8065) and use the `/apps install http http://mattermost-apps-golang-serverless:8080/manifest.json ` slash command to install the webhooks app. Select `Agree to grant the app access to APIs and Locations` and click `Submit` to finish the installation.
```
Is the App a serverless App or a webhooks App? If it's a webhooks App, then I'd be confused as to why it's being referred to in instructions for a serverless App.

azigler replied:
Removed `webhooks`, that was out of place.

neflyte commented on site/content/integrate/apps/quickstart/quick-start-serverless/_index.md:
```diff
@@ -1,436 +1,93 @@
 
-`golang-middleware` template makes use of adding boilerplate handlers to
-`http.DefaultServeMux`, so we just need to add our handlers and serve.
+## Use the app
```
```suggestion
## Use the App
```

neflyte commented on site/content/integrate/apps/quickstart/quick-start-serverless/_index.md:
````diff
@@ -1,436 +1,93 @@
 
-```go
-var deployType apps.DeployType
+You can now use the `/hello-serverless send` command with an optional `message` argument. This will cause the app's bot user to DM you a response:
````
```suggestion
You can now use the `/hello-serverless send` command with an optional `message` argument. This will cause the App's bot user to DM you a response:
```

neflyte commented on site/content/integrate/apps/quickstart/quick-start-serverless/_index.md:
```diff
@@ -1,436 +1,93 @@
 
-	// The main handler for sending a Hello message.
-	http.HandleFunc("/send/submit", send)
+## Review the app
```
```suggestion
## Review the App
```

neflyte commented on site/content/integrate/apps/quickstart/quick-start-serverless/_index.md:
```diff
@@ -1,436 +1,93 @@
 
-	deployType = dt
-}
+To understand the app, examine the following elements:
```
```suggestion
To understand the App, examine the following elements:
```

neflyte commented on site/content/integrate/apps/quickstart/quick-start-serverless/_index.md:
```diff
@@ -1,436 +1,93 @@
-	if ok && v != nil {
-		message += fmt.Sprintf(" ...and %s!", v)
-	}
+The app must provide a manifest, which declares app metadata. In this example, the following permissions are requested via the `/manifest.json` endpoint:
```
```suggestion
The App must provide a manifest, which declares App metadata. In this example, the following permissions are requested via the `/manifest.json` endpoint:
```

neflyte commented on site/content/integrate/apps/quickstart/quick-start-serverless/_index.md:
````diff
@@ -1,436 +1,93 @@
-    lambda.Start(httpadapter.New(http.DefaultServeMux).Proxy)
-}
-```
+Locations are named elements in the Mattermost user interface. Bindings specify how an app's calls should be displayed and invoked from these locations. This app uses only uses the slash command location.
````
```suggestion
Locations are named elements in the Mattermost user interface. Bindings specify how an App's calls should be displayed and invoked from these locations. This App uses only uses the slash command location.
```

neflyte commented on site/content/integrate/apps/quickstart/quick-start-serverless/_index.md:
```diff
@@ -1,436 +1,93 @@
+### Functions
 
-go 1.16
+Functions handle user events and webhooks. The webhooks app uses only one main function via a slash command:
```
```suggestion
Functions handle user events and webhooks. The webhooks App uses only one main function via a slash command:
```

neflyte commented on site/content/integrate/apps/quickstart/quick-start-serverless/_index.md:
```diff
@@ -1,436 +1,93 @@
+### Functions
 
-go 1.16
+Functions handle user events and webhooks. The webhooks app uses only one main function via a slash command:
```
Same concern here about webhooks vs. serverless. It's confusing to see a reference to webhook Apps in a doc about serverless Apps.

azigler replied:
Fixed this.

neflyte commented on site/content/integrate/apps/quickstart/quick-start-serverless/_index.md:
```diff
@@ -1,436 +1,93 @@
-	github.com/mattermost/mattermost-plugin-apps v0.8.0
-	{{path-to-your-module}}/function v0.0.0-00010101000000-000000000000
-)
+- `/hello-serverless send` slash command with an optional `message` argument that sends DMs the user's channel via the app bot.
```
```suggestion
- `/hello-serverless send` slash command with an optional `message` argument that sends DMs the user's channel via the App bot.
```

neflyte commented on site/content/integrate/apps/quickstart/quick-start-serverless/_index.md:
````diff
@@ -1,436 +1,93 @@
 
-replace {{path-to-your-module}}/function => ../function
-```
+All of the logic for the serverless app can be found in `golang/serverless/function/handler.go`, which is the shared function file across all deployment types (HTTP, AWS, and OpenFaaS) for this app. In this file you can review the `send` function that corresponds with the above slash command via a binding. The binding contains a `form` with a `submit` function that sends an interpolated message b ...
````
```suggestion
All of the logic for the serverless App can be found in `golang/serverless/function/handler.go`, which is the shared function file across all deployment types (HTTP, AWS, and OpenFaaS) for this App. In this file you can review the `send` function that corresponds with the above slash command via a binding. The binding contains a `form` with a `submit` function that sends an interpolated message back to the user. As a slash command, the form's one optional `message` will be collected as an argument from the user's command.
```

neflyte commented on site/content/integrate/apps/quickstart/quick-start-serverless/_index.md:
````diff
@@ -1,436 +1,93 @@
-#### `go.mod`
-```
-module {{path-to-your-module}}/http
+Apps may include static assets. Only one asset is used in this example for this app, `icon.png`. Static assets must be served under the `static` path.
````
```suggestion
Apps may include static assets. Only one asset is used in this example for this App, `icon.png`. Static assets must be served under the `static` path.
```

neflyte commented on site/content/integrate/apps/quickstart/quick-start-serverless/_index.md:
```diff
@@ -1,436 +1,93 @@
+Apps may include static assets. Only one asset is used in this example for this app, `icon.png`. Static assets must be served under the `static` path.
 
-go 1.16
+## Uninstall the app
```
```suggestion
## Uninstall the App
```

neflyte commented on site/content/integrate/apps/quickstart/quick-start-serverless/_index.md:
```diff
@@ -1,436 +1,93 @@
-	github.com/mattermost/mattermost-plugin-apps v0.8.0
-	{{path-to-your-module}}/function v0.0.0-00010101000000-000000000000
-)
+Once you're done with the app, you can uninstall it via the `/apps uninstall hello-serverless` slash command. Alternatively, you can use `/apps debug clean` to remove all data for all installed apps.
```
```suggestion
Once you're done with the App, you can uninstall it via the `/apps uninstall hello-serverless` slash command. Alternatively, you can use `/apps debug clean` to remove all data for all installed Apps.
```

neflyte commented on site/content/integrate/apps/quickstart/quick-start-serverless/_index.md:
````diff
@@ -1,436 +1,93 @@
 
-replace {{path-to-your-module}}/function => ../function
-```
+To stop and clean up the app from Docker after you're done, use the following command in the `golang/serverless` directory:
````
```suggestion
To stop and clean up the App from Docker after you're done, use the following command in the `golang/serverless` directory:
```

neflyte commented on site/content/integrate/apps/quickstart/quick-start-serverless/_index.md:
````diff
@@ -1,436 +1,93 @@
-    # docker registry name will be prepended by appsctl
-    image: "hello-openfaas"
+```sh
+docker compose down
````
```suggestion
docker-compose down
```

neflyte commented on site/content/integrate/apps/quickstart/quick-start-serverless/_index.md:
````diff
@@ -1,436 +1,93 @@
-run:
-	go run ./http --manifest=manifest.json --static=static
-```
+Now that you've examined the app in HTTP mode, you can next look at deploying them to your desired serverless plaform. Check out the guides on [how to deploy a Mattermost app to AWS]({{< ref "/integrate/apps/deploy/deploy-aws/" >}}) or [how to deploy a Mattermost app to OpenFaaS]({{< ref "/integrate/apps/deploy/deploy-openfaas/" >}}), depending on which platform you're targeting.
````
```suggestion
Now that you've examined the App in HTTP mode, you can next look at deploying them to your desired serverless plaform. Check out the guides on [how to deploy a Mattermost App to AWS]({{< ref "/integrate/apps/deploy/deploy-aws/" >}}) or [how to deploy a Mattermost App to OpenFaaS]({{< ref "/integrate/apps/deploy/deploy-openfaas/" >}}), depending on which platform you're targeting.
```

neflyte commented on site/content/integrate/apps/quickstart/quick-start-serverless/_index.md:
````diff
@@ -1,436 +1,93 @@
-## dist: creates all bundles
-dist: dist-aws dist-openfaas
-```
+You now know how to create a Mattermost app in Go. If you have questions about building apps or want to show off what you're building, join us on the [Integrations & Apps channel in the Mattermost Community server](https://community.mattermost.com/core/channels/integrations)!
````
```suggestion
You now know how to create a serverless Mattermost App in Go. If you have questions about building Apps or want to show off what you're building, join us on the [Integrations & Apps channel in the Mattermost Community server](https://community.mattermost.com/core/channels/integrations)!
```

azigler commented on site/content/integrate/apps/quickstart/_index.md:
````diff
@@ -1,18 +1,63 @@
+Clone and enter a local copy of the [mattermost-app-examples repository](https://github.com/mattermost/mattermost-app-examples) with the following command:
+
+```sh
+git clone https://github.com/mattermost/mattermost-app-examples && cd mattermost-app-examples
````
I'd like to keep it just for the sake of clarity and ease of copy/paste for the developer.

azigler commented on site/content/integrate/apps/quickstart/_index.md:
```diff
@@ -1,18 +1,63 @@
+
+When you're developing your own app, you need an actual Mattermost server to be running. The development environment accomplishes this by using a [`docker-compose.yml`](https://github.com/mattermost/mattermost-app-examples/blob/master/docker-compose.yml) with a network named `mattermost-apps-dev`. This compose has just two containers: the Mattermost server with a Postgres database. The Mattermost ...
+
+Because they exist on a pre-defined network, other `docker-compose.yml` configurations can connect their containers to this development environment by specifying the same `mattermost-apps-dev` Docker network. Then the Mattermost server can be accessed at `http://mattermost:8065` by any other container on the `mattermost-apps-dev` network. Similarly, your own development apps can then be accessed  ...
```
In regards to your first suggestion, I'm mentioning `docker-compose.yml` again on purpose so the second paragraph is in parallel with the first. I want it to be crystal clear for the reader that we're talking about the same file and format, just in other project directories. I added these other suggestions, though. Thanks for them! 🚀

azigler commented on site/content/integrate/apps/quickstart/_index.md:
````diff
@@ -1,18 +1,63 @@
+Next, use the following command to bring your development environment online.
+
+```sh
+docker compose up -d
````
Looks like you're using an older version of Docker that had a separate docker-compose CLI plugin (e.g., the syntax with `docker-compose`). Refer to [this StackOverflow post](https://stackoverflow.com/a/66526176) and [the Docker docs](https://docs.docker.com/compose/reference/) to see what I mean. `docker compose` is the standard now, the `docker-compose` CLI plugin can be removed.

mickmister replied:
FWIW I use `docker-compose` out of habit, though the CLI yells at me when I do this (or at least it used to?). 0/5 on what should be used here 🤷

azigler commented on site/content/integrate/apps/quickstart/quick-start-go/_index.md:
```diff
@@ -1,288 +1,99 @@
+## Download and start the app
 
-You also need at least `go1.16` installed. Please follow the guide [here](https://golang.org/doc/install) to install the latest version.
+In the same [mattermost-app-examples repository](https://github.com/mattermost/mattermost-app-examples) you cloned via the [developer setup guide]({{< ref "/integrate/apps/quickstart" >}}) above, navigate to the [`golang/hello-world`](https://github.com/mattermost/mattermost-app-examples/tree/master/golang/hello-world) directory and start the Docker compose:
```
I think [stack is a loaded word](https://nickjanetakis.com/blog/docker-tip-23-docker-compose-vs-docker-stack#:~:text=Docker%20Compose%20is%20an%20official,Docker%20containers%20through%20Docker%20Swarm.) in the world of Docker, so I'd like to avoid using it so people don't think `swarm` is involved at all.

azigler commented on site/content/integrate/apps/quickstart/_index.md:
````diff
@@ -1,18 +1,63 @@
+docker compose up -d
+```
+
+Once Mattermost is online, you can access the server at [http://localhost:8065](http://localhost:8065) and create your superuser account. You can also add `127.0.0.1 mattermost` to your `/etc/hosts` file to be able to access Mattermost at [http://mattermost:8065](http://mattermost:8065), which is the value of `MM_SERVICESETTINGS_SITEURL` for Docker networking.
````
This is more like an extra tidbit. It does tell them they can access it at `localhost`, but in some cases you may want to be accessing Mattermost at the same URL that your Docker containers are. For example, image paths would be broken because everything in the server is mapped for `mattermost` as the host name, not `localhost`. I moved it to it's own paragraph and added **Tip:** to try and clarify this is just some extra insight.

azigler commented on site/content/integrate/apps/quickstart/quick-start-jwt/_index.md:
```diff
@@ -0,0 +1,87 @@
+
+## Use the app
+
+You can now use the `/hello-jwt send` command. This will cause the app's bot user to DM you a response:
```
Great idea!

azigler commented on site/content/integrate/apps/quickstart/_index.md:
```diff
@@ -1,18 +1,63 @@
-
-## Tests
+| Golang                  | [Hello World in Go]({{< ref "/integrate/apps/quickstart/quick-start-go" >}})                      |
+| Golang                  | [JWT in Go]({{< ref "/integrate/apps/quickstart/quick-start-jwt" >}})                                   |
```
My thinking is we don't really have another place to put these kinds of guides right now. They're all very introductory looks at how Apps work, and follow the same kind of format and structure. They're also housed in the same repository. I think it makes sense to keep them together, until maybe we have more and it becomes more clear how to divide and reorganize them — and that would be a great problem to have! 🙂

I'd also eventually like to get the guides for lifecycle, oauth2, and webhooks set up in the same way (in future PRs). At that point, it would maybe make more sense to divide them into their own set of guides.

neflyte commented on site/content/integrate/apps/quickstart/quick-start-jwt/_index.md:
```diff
@@ -0,0 +1,85 @@
+
+## Conclusion
+
+You now know how to create a Mattermost App in Go. If you have questions about building Apps or want to show off what you're building, join us on the [Integrations & Apps channel in the Mattermost Community server](https://community.mattermost.com/core/channels/integrations)!
```
```suggestion
You now know how to create a Mattermost App in Go that uses JWTs. If you have questions about building Apps or want to show off what you're building, join us on the [Integrations & Apps channel in the Mattermost Community server](https://community.mattermost.com/core/channels/integrations)!
```

neflyte commented on site/content/integrate/apps/quickstart/quick-start-serverless/_index.md:
````diff
@@ -1,436 +1,93 @@
-## dist: creates all bundles
-dist: dist-aws dist-openfaas
-```
+You now know how to create a Mattermost App in Go. If you have questions about building Apps or want to show off what you're building, join us on the [Integrations & Apps channel in the Mattermost Community server](https://community.mattermost.com/core/channels/integrations)!
````
```suggestion
You now know how to create a Mattermost App in Go that deploys in serverless environments. If you have questions about building Apps or want to show off what you're building, join us on the [Integrations & Apps channel in the Mattermost Community server](https://community.mattermost.com/core/channels/integrations)!
```

neflyte reviewed (changes requested):
Looks awesome, @azigler! :smile: 
I found a few (not 150) more things to consider, but it's looking leaps and bounds better!

azigler commented on site/content/integrate/apps/quickstart/quick-start-js/_index.md:
```diff
@@ -1,265 +1,89 @@
+Before you can start with your App, you first need to set up your environment by following the [developer setup guide]({{< ref "/integrate/apps/quickstart" >}}).
 
-## Build the App
+You also need an [LTS version of NodeJS](https://nodejs.org/en/download/).
```
You're looking at what will soon be the deprecated `dev` environment and example. The new one is here and uses Node Gallium (v16) LTS https://github.com/mattermost/mattermost-app-examples/blob/master/typescript/hello-world/docker-compose.yml. I will specify here that they need Node v16 LTS.

neflyte commented:
Regarding page weights: the weight of a page determines its order in the LHS. If a weight is not defined then the page is sorted alphabetically. A lower weight value means a higher position in the LHS. 

Note that if a directory contains an `_index.md` file, the weight defined in that file applies to the topic in the **parent directory**. Weights in other `.md` files in the same directory are used as normal and the pages are treated as sub-topics of the parent.

---

Regarding renaming a directory: it's perfectly fine to rename a directory, however page redirects need to be managed. Page redirects are defined in the `aliases` key of the page frontmatter, the YAML section at the top of each `.md` file. Each element of the `aliases` list is transformed to a physical HTML file that contains a redirect to the page you're modifying.

For example, if we wanted to rename the directory `/integrate/apps/quickstart/quick-start-js` to `/integrate/apps/quickstart/quick-start-ts`, we would:
- Add an element to the `aliases` list in the page frontmatter with the current (undesired) path:
```yaml
---
# ...
aliases:
  - /integrate/apps/quickstart/quick-start-js
# ...
---
```
- Rename the directory from `quick-start-js` to `quick-start-ts`

I hope this helps! :smile:

azigler commented:
> docker-compose is not a plugin in the Docker installations I've used over the years; it's a separate program altogether. I also wasn't aware that Docker made this change; they didn't email me about it and I'm subscribed to their mailing list.
>
> Are we suggesting that the reader has "Docker Desktop" installed rather than straight Docker? The Docker that I install on Ubuntu using apt doesn't seem to be "Docker Desktop". We should be explicit about what release of Docker should be used by adding a note section here. There are many developers who will not have known about the changes to docker-compose and will hit errors when trying to run the commands we're documenting.

"Docker Desktop" or "Docker Engine" both work, if they follow the install instructions for whichever one they pick. I will clarify in the instructions.

If you do `docker compose version` and `docker-compose version` you'll get the same version, because it's aliased. If `docker compose` doesn't work for you, then something is wrong with your configuration and I recommend following the installation steps again. As for the `docker -v` output, on my remote Linux with Docker Engine (the CLI), it's `Docker version 20.10.17, build 100c701`. On my Mac with Docker Desktop (the application), it's also `Docker version 20.10.17, build 100c701` because they're the same under the hood. If someone installs Docker today (either application type) or updates to a modern install (either application type), they get the version with `compose` as a built-in alias command. [They are planning to have this alias in place indefinitely](https://www.docker.com/blog/announcing-compose-v2-general-availability/). `docker compose` is the new standard way the command is written.

> I missed one. There's a typo in the heading of the /integrate/apps/quickstart/quick-start-js page. It says Create a wello world Mattermost App in TypeScript instead of Create a hello world Mattermost App in TypeScript.

Hah, fixed! Thanks.

mickmister commented on site/content/integrate/apps/_index.md:
```diff
@@ -45,7 +45,7 @@ The built-in [incoming webhook]({{< ref "/integrate/webhooks/incoming" >}}) is g
 A plugin should be used when you need to [directly alter the UI]({{< ref "integrate/plugins/components/webapp/best-practices" >}}) in Mattermost or you have a feature that requires low latency with the server (such as replacing characters in any message [before it is saved]({{< ref "integrate/plugins/components/server/reference#Hooks.MessageWillBePosted" >}})). Currently, plugins have several [UX ...
 
-### What's the difference between the Apps framework and the plugin framework?
+### What's the difference between the Apps Framework and the plugin framework?
```
We have a lot of inconsistencies across the board with things like this:

- App Framework
- App framework
- Apps Framework
- Apps framework
- apps framework
- App Bar (People seldom refer to it as this, though this is what the code refers to it as. It's called an "egg carton", not an "eggs carton" :wink:)
- Apps Bar
- Apps bar
- apps bar

We should converge on the spelling/naming of these things, and have a reference doc to use when there is a question about this. There are often inconsistencies in the same paragraph mentioning these things. I'm not sure who the arbiter of these naming decisions should be but we absolutely need to make a decision on this so we can be consistent and have credibility in our docs and blog posts.

azigler replied:
Agreed @mickmister. This is a problem all across the docs, not just isolated to this PR. I am getting conflicting information on capitalizing and we really need a reference doc. cc @cwarnermm

cwarnermm replied:
Agreed that this is a great place to standardize for consistency and clarity, and that this is an issue across all Mattermost doc sets. @justinegeffen is in the process of pulling together the many terminology decisions, including capitalization, that have been made over the course of this year into a reference resource. @justinegeffen - this is another great candidate worth including.

My own two cents: I propose that we use lowercase (apps framework) when referring to the app framework technology we're developing across the doc sets. I don't see it as a first-class citizen or product that requires capitalization.

Any concerns about this proposed approach?

mickmister commented on site/content/integrate/apps/quickstart/quick-start-go/_index.md:
```diff
@@ -1,288 +1,99 @@
-heading: "Write a Mattermost app in Go"
-description: "This quick start guide will walk you through the basics of writing a Mattermost App in Go."
+title: "Hello World in Go"
+heading: "Create a Hello World Mattermost app in Go"
```
Maybe:

Create a "Hello World" Mattermost App in Go

mickmister commented on site/content/integrate/apps/_index.md:
```diff
@@ -33,53 +33,53 @@ When you develop using the Apps Framework, your apps can:
 That depends on your use case, as they each have benefits.
 
-The built-in [incoming webhook]({{< ref "/integrate/webhooks/incoming" >}}) is great for simple use cases. It requires the incoming payload to contain a valid {{< newtabref href="https://pkg.go.dev/github.com/nhannv/mattermost-server/model#Post" title="Post" >}} JSON object, but some external systems don't allow customization of the fields included in the webhook payload. Compared to built-in web ...
+The built-in [incoming webhook]({{< ref "/integrate/webhooks/incoming" >}}) is great for simple use cases. It requires the incoming payload to contain a valid {{< newtabref href="https://pkg.go.dev/github.com/mattermost/mattermost-server/v6/model#Post" title="Post" >}} JSON object, but some external systems don't allow customization of the fields included in the webhook payload. Compared to built ...
```
I (selfishly) like the way I worded this [here](https://community.mattermost.com/core/pl/4jtbfepncigyfxxfnfnycbfn1c)

> Each type of existing integration type is sort of a la carte, whereas the App framework is an all-in-one package that supports its own version of some of those out of the box. Meaning when you install an App, you don't need to go create a separate bot account, OAuth app, webhooks etc. Some pieces are reused, such as the bot account mechanism, but some pieces were remade and are not reused, such as webhooks.
>
> In the case of webhooks, the existing mechanism is only able to create posts, and only accepts a webhook payload that matches a Mattermost post's data structure. It does not support other logic for handling arbitrary data structures from external systems.

Just some additional context we can provide here. Also, should we have a table of contents on this page so this section is easier to find? Or maybe have its own more discoverable page, that is agnostic to a specific integration type?

azigler replied:
I remember this wording -- thanks for bringing this back up. I've incorporated it here. Like I mentioned in my other comment, I think a revamp of this page would be a future PR. There's a lot to sort out here and we should be more clear on the goals of the page.

mickmister commented on site/content/integrate/apps/_index.md:
```diff
@@ -11,11 +11,11 @@ Apps are lightweight, interactive add-ons to Mattermost which can be hosted as H
 - Include dynamic fields, code as an on-demand function, and interactive modals.
 
-When you develop using the Apps Framework, your apps can:
+When you develop using the App framework, your Apps can:
```
There's an inconsistency on this page for "App framework" vs "Apps framework"

mickmister commented on site/content/integrate/apps/_index.md:
```diff
@@ -33,53 +33,53 @@ When you develop using the Apps Framework, your apps can:
+The built-in [incoming webhook]({{< ref "/integrate/webhooks/incoming" >}}) is great for simple use cases. It requires the incoming payload to contain a valid {{< newtabref href="https://pkg.go.dev/github.com/mattermost/mattermost-server/v6/model#Post" title="Post" >}} JSON object, but some external systems don't allow customization of the fields included in the webhook payload. Compared to built ...
 
-A plugin should be used when you need to [directly alter the UI]({{< ref "integrate/plugins/components/webapp/best-practices" >}}) in Mattermost or you have a feature that requires low latency with the server (such as replacing characters in any message [before it is saved]({{< ref "integrate/plugins/components/server/reference#Hooks.MessageWillBePosted" >}})). Currently, plugins have several [UX ...
+A plugin should be used when you need to [directly alter the UI]({{< ref "integrate/plugins/components/webapp/best-practices" >}}) in Mattermost or you have a feature that requires low latency with the server (such as replacing characters in any message [before it is saved]({{< ref "integrate/plugins/components/server/reference#Hooks.MessageWillBePosted" >}})). Currently, plugins have several [UX ...
```
> however we plan to add/migrate more UX hooks into the Apps framework over time

We more-or-less have all of the hooks we want to support there. Not sure if we are going to add more. If we do, they will likely be specialized/enhanced version of the plugin hooks, like the "Bot Mentioned" hook.

mickmister commented on site/content/integrate/apps/_index.md:
```diff
@@ -33,53 +33,53 @@ When you develop using the Apps Framework, your apps can:
-### What language should I use to write apps?
+### What language should I use to write Apps?
 
 Any language you want. We currently have an [official driver for Go](https://pkg.go.dev/github.com/mattermost/mattermost-plugin-apps/apps), and we are planning to release other drivers in the future in conjunction with our community.
```
Not sure if we are going to add more drivers

```suggestion
Any language you want. We currently have an [official driver for Go](https://pkg.go.dev/github.com/mattermost/mattermost-plugin-apps/apps).
```

mickmister commented on site/content/integrate/apps/_index.md:
```diff
@@ -33,53 +33,53 @@ When you develop using the Apps Framework, your apps can:
+   You can find more information about the Marketplace [here](https://mattermost.com/marketplace/). Example AWS Lambda Apps can be
    found
    [here](https://github.com/mattermost/mattermost-plugin-apps/tree/master/examples/js/aws_hello)
    (JavaScript) and
```
```suggestion
   (TypeScript) and
```

mickmister commented on site/content/integrate/apps/_index.md:
```diff
@@ -33,53 +33,53 @@ When you develop using the Apps Framework, your apps can:
    [OpenFaaS](https://www.openfaas.com/), and [Kubeless](https://kubeless.io/)
    are supported, with plans for more serverless platforms, Kubernetes, and
-   docker compose. Customer-deployable apps are not yet distributed via the
+   docker compose. Customer-deployable Apps are not yet distributed via the
```
To me, deploying via docker compose is the same as deploying an HTTP server. Each language/environment is different and requires a different setup for deployment outside of the automated lambda function strategy. Not sure if anything needs to change here though.

mickmister commented on site/content/integrate/apps/quickstart/quick-start-jwt/_index.md:
```diff
@@ -0,0 +1,85 @@
+weight: 30
+---
+
+This quick start guide will walk you through the basics of using [JSON Web Tokens (JWT)](https://jwt.io/) in a Mattermost App in Go. In this guide you will review an App that:
```
I think we should explain why you would want to use a JWT token for App development. It's so the App can validate the legitimacy of the request, to verify that this request is indeed coming from the Mattermost server we are expecting it to come from.

azigler replied:
Great point, I worked this in!

mickmister commented on site/content/integrate/apps/quickstart/quick-start-jwt/_index.md:
````diff
@@ -0,0 +1,85 @@
+
+```
+Use '/apps install http http://mattermost-apps-golang-jwt:8084/manifest.json' to install the app
+Use "1234" as the app's JWT secret
````
Maybe point out that this secret is something that you should generate at random per App installation in production, and store on your end to identify the specific Mattermost server.

azigler replied:
Excellent point, added this.

mickmister commented on site/content/integrate/apps/quickstart/quick-start-jwt/_index.md:
```diff
@@ -0,0 +1,85 @@
+
+## Install the App on Mattermost
+
+Next, access your development Mattermost Server at [http://localhost:8065](http://localhost:8065) and use the `/apps install http http://mattermost-apps-golang-jwt:8084/manifest.json` slash command to install the JWT App. Select `Agree to grant the app access to APIs and Locations`, enter the JWT secret (1234), and click `Submit` to finish the installation.
```
I personally think we should spell it "Mattermost server", except for when we are talking directly about the product like "the Mattermost Server". 2/5 An instance a Mattermost server doesn't need to be capitalized.

azigler replied:
I'll gladly go through and standardize these guides if we can establish a reference document, or someone can do the whole repo in one big sweep (might be simpler). I agree in this case it probably doesn't ned to be capitalized, I had it lowercase before.

cwarnermm replied:
Another great consistency point to clarity! The product docs use a capital "S" when referring to the binary, and a lowercase "s" when referring to the system that's running Mattermost code.

cwarnermm replied:
@azigler - I'd recommend holding off on an end-to-end sweep until @justinegeffen makes the terminology reference available.

mickmister reviewed (commented):
Awesome work @azigler!! I have lots of comments on here. Some of them apply to multiple pages due to similar content existing on different pages. Overall this is looking great! :rocket: Thanks for your work on this!!

azigler commented on site/content/integrate/apps/quickstart/_index.md:
```diff
@@ -1,18 +1,63 @@
+
+When you're developing your own App, you need an actual Mattermost Server to be running. The development environment accomplishes this by using a [`docker-compose.yml`](https://github.com/mattermost/mattermost-app-examples/blob/master/docker-compose.yml) with a network named `mattermost-apps-dev`. This Docker Compose file has just two containers: the Mattermost Server and a Postgres database. The ...
+
+Because they exist on a pre-defined network, other `docker-compose.yml` configurations can connect their containers to this development environment by specifying the same `mattermost-apps-dev` Docker network. Then the Mattermost Server can be accessed at `http://mattermost:8065` by any other container on the `mattermost-apps-dev` network. Similarly, your own development Apps can then be accessed  ...
```
Great idea, I will incorporate this!

azigler commented on site/content/integrate/apps/quickstart/_index.md:
```diff
@@ -1,18 +1,63 @@
+
+Once Mattermost is online, you can access the server at [http://localhost:8065](http://localhost:8065) and create your superuser account.
+
+**Tip:** You can also add `127.0.0.1 mattermost` to your `/etc/hosts` file to be able to access Mattermost at [http://mattermost:8065](http://mattermost:8065), which is the value of `MM_SERVICESETTINGS_SITEURL` for Docker networking.
```
Great idea, adding this tidbit.

azigler commented on site/content/integrate/apps/_index.md:
```diff
@@ -33,53 +33,53 @@ When you develop using the Apps Framework, your apps can:
 Apps are now generally available and we always appreciate your feedback! Share constructive feedback in the [Mattermost Apps channel](https://community.mattermost.com/core/channels/mattermost-apps) on our Mattermost community instance.
 
-Read the [JavaScript start guide]({{< ref "quickstart/quick-start-js" >}}) or [Go quick start guide]({{< ref "quickstart/quick-start-go" >}}) to learn how to write your first App.
+Read the quick start guides for [Hello World in TypeScript]({{< ref "quickstart/quick-start-ts" >}}) or [Hello World in Go]({{< ref "quickstart/quick-start-go" >}}) to learn how to write your first App.
```
Excellent suggestion.

azigler commented on site/content/integrate/apps/quickstart/quick-start-go/_index.md:
```diff
@@ -1,288 +1,99 @@
 ## Prerequisites
 
-Before you can start with your app, you first need to set up a local developer environment following the [server]({{< ref "/contribute/server/developer-setup" >}}) and [webapp]({{< ref "/contribute/webapp/developer-setup" >}}) setup guides. You must enable the apps feature flag before starting the Mattermost server by setting the environment variable `MM_FEATUREFLAGS_AppsEnabled` to `true` by e.g ...
+Before you can start with your App, you first need to set up your environment by following the [developer setup guide]({{< ref "/integrate/apps/quickstart" >}}).
```
Great point, I'll fix that language across the guides.

azigler commented on site/content/integrate/apps/quickstart/quick-start-go/_index.md:
````diff
@@ -1,288 +1,99 @@
 
-```go
-package main
+![image](modal.png)
````
Okay, I'll change that in the examples repo.

azigler commented on site/content/integrate/apps/_index.md:
```diff
@@ -11,11 +11,11 @@ Apps are lightweight, interactive add-ons to Mattermost which can be hosted as H
-- Work on both Mobile and Desktop clients so developers can focus on the functionality of their apps.
+- Work on both Mobile and Desktop clients so developers can focus on the functionality of their Apps.
 - Be deployed using our serverless hosting infrastructure keeping data secure and supporting scalability by being stateless.
 - Include dynamic fields, code as an on-demand function, and interactive modals.
```
I agree, but this page is mostly outside the scope of my PR. The only changes I'm trying to make this page are the things that point to the quick start guides. I think this page need a separate facelift. 2/5 on this.

azigler commented on site/content/integrate/apps/quickstart/quick-start-ts/_index.md:
```diff
@@ -0,0 +1,89 @@
+
+Before you can start with your App, you first need to set up your environment by following the [developer setup guide]({{< ref "/integrate/apps/quickstart" >}}).
+
+You also need [NodeJS v16 (Gallium) LTS](https://nodejs.org/en/download/).
```
I don't know what range of versions the App framework works with. [Gallium is the current Active LTS for Node](https://nodejs.org/en/about/releases/) so I always default to it. @neflyte pointed out that we don't want to just say "LTS" because there are older LTS versions that probably don't work.

mickmister replied:
@azigler I still think we should de-emphasize and just say:

> Download the LTS or current version of Node [here](https://nodejs.org/en)

That link gives only the latest relevant versions for Node.

My concern is that many people will think they need that exact version of Node that is currently mentioned above. Some will upgrade what they currently have, and some will downgrade what they currently have, when in reality both scenarios could have been using a valid version that works with our software here.

Saying "download this exact version of x" can be seen as a red flag of compatibility, and really just gives me frustration as a dev that I need to configure something in a specific way for something I'm working on. We already have this actual issue with plugin projects so I know the burden of this, and having to explain to individual people that they need to use a specific version of Node kind of sucks for both parties (and I of course feel the judgment of everyone watching the conversation about _having_ to explain that, due to non-updated projects our team owns).

In the case of the App framework docs, I think we should be vague about this version, since we don't have a reason to say why any version is not compatible. As long as our JS libraries (which are pretty minimal) are compatible with what's in that link, it should work fine.

neflyte replied:
I suggest we select a minimum-compatible version of NodeJS and stick with it. If that's Node v14 or v16, it doesn't matter; we can always update this in the future.

The downside of not selecting a version - but not necessarily a point release - is developers who run into dependency issues and then have to reach out for assistance. I realize avoiding dependency issues in JS is fraught with peril, however I feel we should do the best we can to cushion implementers against issues that we know about.

azigler commented on site/content/integrate/apps/quickstart/quick-start-go/_index.md:
````diff
@@ -1,288 +1,99 @@
-```
-/apps uninstall hello-world
-```
+You now know how to create a Mattermost App in Go that can deploy to serverless environments. If you have questions about building Apps or want to show off what you're building, join us on the [Integrations & Apps channel in the Mattermost Community server](https://community.mattermost.com/core/channels/integrations)!
````
Yeah, this shouldn't be here. Thanks!

azigler commented on site/content/integrate/apps/quickstart/quick-start-go/_index.md:
```diff
@@ -1,288 +1,99 @@
-		}
-	}
-}
+Use '/apps install http http://mattermost-apps-golang-hello-world:4000/manifest.json' to install the App
```
Makes sense, I'll move these all to blockquotes.

azigler commented on site/content/integrate/apps/quickstart/quick-start-go/_index.md:
```diff
@@ -1,288 +1,99 @@
-	fmt.Println("Use '/apps install http http://localhost" + addr + "/manifest.json' to install the app") // matches manifest.json
-	log.Fatal(http.ListenAndServe(addr, nil))
-}
+### Functions and forms
```
I agree, I made some tweaks across the guides to make it look less staunch and special.

azigler commented on site/content/integrate/apps/quickstart/quick-start-go/_index.md:
```diff
@@ -1,288 +1,99 @@
-	// The form for sending a Hello message.
-	http.HandleFunc("/send/form", httputils.DoHandleJSONData(formData))
+- Create posts as a bot.
+- Render icons in the channel header that will communicate with your App when clicked.
```
Okay, I added that to the `docker-compose.yml` in the examples repo. Will that resolve this?

mickmister replied:
Yep sounds great :+1:

azigler commented on site/content/integrate/apps/quickstart/quick-start-python/_index.md:
```diff
@@ -0,0 +1,91 @@
+
+The App has to provide a manifest, which declares the App's metadata required for installation. In this example, the following permissions and locations are requested:
+
+- Create posts as a bot (`act_as_bot`)
+- Render icons in the channel header (`/channel_header`)
+- Create slash commands (`/command`)
```
Standardized this one, because TS was the same better way.

azigler commented on site/content/integrate/apps/_index.md:
```diff
@@ -11,11 +11,11 @@ Apps are lightweight, interactive add-ons to Mattermost which can be hosted as H
 - Attach themselves to locations in the Mattermost user interface (e.g. channel bar buttons, post menu, channel menu, commands), and can add their custom slash commands with full autocomplete.
 - Receive webhooks from Mattermost, and third-parties, and use the Mattermost REST APIs to post messages, etc.
-- Work on both Mobile and Desktop clients so developers can focus on the functionality of their apps.
+- Work on both Mobile and Desktop clients so developers can focus on the functionality of their Apps.
```
I like this rewording. The Apps can _and_ do, I think it works out.

azigler commented:
> Awesome work @azigler!! I have lots of comments on here. Some of them apply to multiple pages due to similar content existing on different pages. Overall this is looking great! 🚀 Thanks for your work on this!!

Thanks @mickmister for your review! I incorporated most of your suggestions and left some comments.

mickmister reviewed (approved):
:tada:

azigler commented on site/content/integrate/apps/_index.md:
```diff
@@ -33,17 +34,19 @@ When you develop using the App framework, your Apps can:
-The built-in [incoming webhook]({{< ref "/integrate/webhooks/incoming" >}}) is great for simple use cases. It requires the incoming payload to contain a valid {{< newtabref href="https://pkg.go.dev/github.com/mattermost/mattermost-server/v6/model#Post" title="Post" >}} JSON object, but some external systems don't allow customization of the fields included in the webhook payload. Compared to built ...
+Each type of existing integration type is sort of a la carte, whereas the App framework is an all-in-one package that supports its own version of some of those out of the box. Meaning when you install an App, you don't need to go create a separate bot account, OAuth app, webhooks etc. Some pieces are reused, such as the bot account mechanism, but some pieces were remade and are not reused, such a ...
+
+In the case of webhooks, the existing mechanism is only able to create posts, and only accepts an [incoming webhook]({{< ref "/integrate/webhooks/incoming" >}}) payload that matches a Mattermost post's data structure. It does not support other logic for handling arbitrary data structures from external systems.
```
Knowing that, how would you recommend this be reworded?

mickmister replied:
Maybe just remove the "that matches a Mattermost post's data structure" fragment

levb reviewed (approved):
🎉 📜 👍

mickmister commented on site/content/integrate/apps/quickstart/_index.md:
```diff
@@ -28,6 +28,8 @@ You can also access Mattermost from outside the Docker network via [http://local
 
 To change the Mattermost Apps plugin or the Mattermost Server versions, you can edit [`docker-compose.yml`](https://github.com/mattermost/mattermost-app-examples/blob/master/docker-compose.yml).
 
+Be sure to create a `.docker.env` file in the same directory as the `docker-compose.yml` file to store any secrets. Or you can rename and use [`.docker.env.example`](https://github.com/mattermost/mattermost-app-examples/blob/master/.docker.env.example) in the repository to see some example secrets you may use while building Apps.
```
This is worded in a way that makes me think "in case you need to store secrets, you'll need to create this file", when in reality, it's required to make this run. Feel free to reword parts of the suggestion below:

```suggestion
You'll need to create a `.docker.env` file in the same directory as the `docker-compose.yml` file in order to complete the Docker Compose configuration. Use this file to store any secrets or any other environment variables you wish to use with the Mattermost server. Check out the [`.docker.env.example`](https://github.com/mattermost/mattermost-app-examples/blob/master/.docker.env.example) file in the examples repository to see some secrets you may want to use while building Apps. The `.docker.env` file is ignored in the repository's `.gitignore` file, so these secrets won't be committed to the repository if you make any edits.
```

azigler replied:
Agreed! Great suggestion.

neflyte commented on site/content/integrate/apps/_index.md:
```diff
@@ -5,23 +5,24 @@ description: "Apps are lightweight interactive add-ons to Mattermost."
   ![image](app-slash-command-zoomed-in.png)
 
-- Post messages to channels
+- Post messages to channels (and anything else possible with [Mattermost's API](https://api.mattermost.com/))
```
```suggestion
- Post messages to channels (and anything else possible with {{<newtabref title="Mattermost's API" href="https://api.mattermost.com/">}})
```

neflyte commented on site/content/integrate/apps/_index.md:
```diff
@@ -33,17 +34,19 @@ When you develop using the App framework, your Apps can:
 That depends on your use case, as they each have benefits.
 
-The built-in [incoming webhook]({{< ref "/integrate/webhooks/incoming" >}}) is great for simple use cases. It requires the incoming payload to contain a valid {{< newtabref href="https://pkg.go.dev/github.com/nhannv/mattermost-server/model#Post" title="Post" >}} JSON object, but some external systems don't allow customization of the fields included in the webhook payload. Compared to built-in web ...
+Each type of existing integration type is sort of a la carte, whereas the App framework is an all-in-one package that supports its own version of some of those out of the box. Meaning when you install an App, you don't need to go create a separate bot account, OAuth app, webhooks etc. Some pieces are reused, such as the bot account mechanism, but some pieces were remade and are not reused, such a ...
```
```suggestion
Each type of existing integration is sort of a la carte, whereas the App framework is an all-in-one package that supports its own version of some of those out of the box. Meaning when you install an App, you don't need to go create a separate bot account, OAuth app, webhooks etc. Some pieces are reused, such as the bot account mechanism, but some pieces were remade and are not reused, such as webhooks.
```

neflyte replied:
```suggestion
Each type of existing integration is sort of a la carte, whereas the App framework is an all-in-one package that supports its own version of some of those out of the box. Meaning when you install an App, you don't need to go create a separate bot account, OAuth app, webhooks etc. Some pieces are reused, such as the bot account mechanism, but some pieces were remade and are not reused, such as webhooks.
```

neflyte commented on site/content/integrate/apps/_index.md:
```diff
@@ -5,23 +5,24 @@ description: "Apps are lightweight interactive add-ons to Mattermost."
   ![image](app-slash-command-zoomed-in.png)
 
-- Post messages to channels
+- Post messages to channels (and anything else possible with [Mattermost's API](https://api.mattermost.com/))
```
```suggestion
- Post messages to channels (and anything else possible with {{<newtabref title="Mattermost's API" href="https://api.mattermost.com/">}})
```

This suggestion will make the link open in a new tab. We should do this for any link that takes the user away from the docs site.

cwarnermm replied:
I like this approach!

neflyte commented on site/content/integrate/apps/_index.md:
```diff
@@ -33,17 +34,19 @@ When you develop using the App framework, your Apps can:
+In the case of webhooks, the existing mechanism is only able to create posts, and only accepts an [incoming webhook]({{< ref "/integrate/webhooks/incoming" >}}) payload. It does not support other logic for handling arbitrary data structures from external systems.
 
-A plugin should be used when you need to [directly alter the UI]({{< ref "integrate/plugins/components/webapp/best-practices" >}}) in Mattermost or you have a feature that requires low latency with the server (such as replacing characters in any message [before it is saved]({{< ref "integrate/plugins/components/server/reference#Hooks.MessageWillBePosted" >}})). Currently, plugins have several [UX ...
+A plugin should be used when you need to [directly alter the UI]({{< ref "integrate/plugins/components/webapp/best-practices" >}}) in Mattermost or you have a feature that requires low latency with the server (such as replacing characters in any message [before it is saved]({{< ref "integrate/plugins/components/server/reference#Hooks.MessageWillBePosted" >}})). It should be noted that plugins hav ...
```
```suggestion
A plugin should be used when you need to [directly alter the UI]({{< ref "integrate/plugins/components/webapp/best-practices" >}}) in Mattermost or you have a feature that requires low latency with the server (such as replacing characters in any message [before it is saved]({{< ref "integrate/plugins/components/server/reference#Hooks.MessageWillBePosted" >}})).

{{<note "Note:">}}
Plugins have several [UX hooks]({{< ref "integrate/plugins/components/server/reference#Hooks" >}}) that Apps cannot access. Please see the [plugin documentation]({{< ref "integrate/plugins" >}}) for more information.
{{</note>}}
```

neflyte commented on site/content/integrate/apps/_index.md:
```diff
@@ -54,7 +57,7 @@ The Apps framework provides a few differences from plugins, including:
 ### What language should I use to write Apps?
 
-Any language you want. We currently have an [official driver for Go](https://pkg.go.dev/github.com/mattermost/mattermost-plugin-apps/apps), and we are planning to release other drivers in the future in conjunction with our community.
+Any language you want. We currently have an [official driver for Go](https://pkg.go.dev/github.com/mattermost/mattermost-plugin-apps/apps).
```
```suggestion
Any language you want. We currently have an {{<newtabref title="official driver for Go" href="https://pkg.go.dev/github.com/mattermost/mattermost-plugin-apps/apps">}}.
```

neflyte commented on site/content/integrate/apps/_index.md:
```diff
@@ -63,10 +66,8 @@ Any language you want. We currently have an [official driver for Go](https://pkg
-   (JavaScript) and
-   [here](https://github.com/mattermost/mattermost-plugin-apps/tree/master/examples/go/hello-serverless)
-   (go).
+   [here](https://github.com/mattermost/mattermost-app-examples/tree/master/golang/serverless)
```
```suggestion
   {{<newtabref title="here" href="https://github.com/mattermost/mattermost-app-examples/tree/master/golang/serverless">}}
```

neflyte commented on site/content/integrate/apps/quickstart/_index.md:
```diff
@@ -18,12 +18,18 @@ Clone and enter a local copy of the [mattermost-app-examples repository](https:/
-When you're developing your own App, you need an actual Mattermost Server to be running. The development environment accomplishes this by using a [`docker-compose.yml`](https://github.com/mattermost/mattermost-app-examples/blob/master/docker-compose.yml) with a network named `mattermost-apps-dev`. This compose has just two containers: the Mattermost Server with a Postgres database. The Mattermost ...
+When you're developing your own App, you need an actual Mattermost Server to be running. The development environment accomplishes this by using a [`docker-compose.yml`](https://github.com/mattermost/mattermost-app-examples/blob/master/docker-compose.yml) with a network named `mattermost-apps-dev`. This Docker Compose file has just two containers: the Mattermost Server and a Postgres database.
+
+The Mattermost image itself is preconfigured with settings for local App development but you can customize them. For example, the `MM_SERVICESETTINGS_SITEURL` environment variable is set to `http://mattermost:8065` on the Mattermost container, but you could change this to be any URL, like one from [ngrok](https://ngrok.com/) or [Gitpod](https://gitpod.io/).
```
```suggestion
The Mattermost image itself is preconfigured with settings for local App development but you can customize them. For example, the `MM_SERVICESETTINGS_SITEURL` environment variable is set to `http://mattermost:8065` on the Mattermost container, but you could change this to be any URL, like one from {{<newtabref title="ngrok" href="https://ngrok.com/">}} or {{<newtabref title="Gitpod" href="https://gitpod.io/">}}.
```

neflyte commented on site/content/integrate/apps/quickstart/_index.md:
```diff
@@ -18,12 +18,18 @@ Clone and enter a local copy of the [mattermost-app-examples repository](https:/
+You'll need to create a `.docker.env` file in the same directory as the `docker-compose.yml` file in order to complete the Docker Compose configuration. Use this file to store any secrets or any other environment variables you wish to use with the Mattermost server. Check out the [`.docker.env.example`](https://github.com/mattermost/mattermost-app-examples/blob/master/.docker.env.example) file in ...
+
 Next, use the following command to bring your development environment online.
 
```
The commands below will fail with Docker Engine. Please add notes where necessary to direct the user to the command that works for Docker Engine.

Otherwise, please remove all references to Docker Engine and add a note that only Docker Desktop is supported for developing Apps.

azigler replied:
Okay, I'll add a note about using the hyphenated commands if you're on DE with the compose plguin.

neflyte commented on site/content/integrate/apps/quickstart/quick-start-go/_index.md:
```diff
@@ -96,4 +94,4 @@ docker compose down
 ## Conclusion
 
-You now know how to create a Mattermost App in Go. If you have questions about building apps or want to show off what you're building, join us on the [Integrations & Apps channel in the Mattermost Community server](https://community.mattermost.com/core/channels/integrations)!
+You now know how to create a Mattermost App in Go. If you have questions about building Apps or want to show off what you're building, join us on the [Mattermost Apps channel in the Mattermost Community server](https://community.mattermost.com/core/channels/mattermost-apps)!
```
```suggestion
You now know how to create a Mattermost App in Go. If you have questions about building Apps or want to show off what you're building, join us on the {{<newtabref title="Mattermost Apps" href="https://community.mattermost.com/core/channels/mattermost-apps">}} channel in the Mattermost Community server!
```

neflyte commented on site/content/integrate/apps/quickstart/quick-start-jwt/_index.md:
```diff
@@ -2,17 +2,19 @@
+- Contains a `manifest.json`, declares itself an HTTP application that uses a bot account, uses JWT, and attaches UI elements to locations in the Mattermost interface.
 - Attaches the `send` function to a `/hello-jwt` command and wraps it with a function to authenticate JWT.
 
+Why you would want to use JWT for App development? Using a token, your App can validate the legitimacy of any request verify that this request is indeed coming from the Mattermost server we are expecting it to come from. Simply put, implementing JWT will enhance your App's security and functionality.
```
This doesn't read quite right. I think there's a comma missing before `verify`.

azigler replied:
I'll fix this one.

neflyte commented on site/content/integrate/apps/quickstart/quick-start-jwt/_index.md:
```diff
@@ -82,4 +90,4 @@ docker compose down
 ## Conclusion
 
-You now know how to create a Mattermost App in Go. If you have questions about building Apps or want to show off what you're building, join us on the [Integrations & Apps channel in the Mattermost Community server](https://community.mattermost.com/core/channels/integrations)!
+You now know how to create a Mattermost App in Go that uses JWTs. If you have questions about building Apps or want to show off what you're building, join us on the [Mattermost Apps channel in the Mattermost Community server](https://community.mattermost.com/core/channels/mattermost-apps)!
```
```suggestion
You now know how to create a Mattermost App in Go that uses JWTs. If you have questions about building Apps or want to show off what you're building, join us on the {{<newtabref title="Mattermost Apps" href="https://community.mattermost.com/core/channels/mattermost-apps">}} channel in the Mattermost Community server!
```

neflyte commented on site/content/integrate/apps/quickstart/quick-start-python/_index.md:
```diff
@@ -66,11 +66,11 @@ The App has to provide a manifest, which declares the App's metadata required fo
 ### Bindings and locations
 
-This App adds a channel header button and slash commands. In order to register these locations, there is a `POST` handler for the `/bindings` endpoint in your App's API.
+[Bindings]({{< ref "/integrate/apps/structure/bindings" >}}) specify how an App's calls should be displayed and invoked from these locations. This App adds a channel header button and slash commands. In order to register these locations, there is a `POST` handler for the `/bindings` endpoint in your App's API.
```
What locations are being referred to by the phrase "these locations"?
It doesn't look like locations were introduced before this instance.
Could we add some text about what locations are?

neflyte commented on site/content/integrate/apps/quickstart/quick-start-python/_index.md:
```diff
@@ -88,4 +88,4 @@ docker compose down
 ## Conclusion
 
-You now know how to create a Mattermost App in Python. If you have questions about building Apps or want to show off what you're building, join us on the [Integrations & Apps channel in the Mattermost Community server](https://community.mattermost.com/core/channels/integrations)!
+You now know how to create a Mattermost App in Python. If you have questions about building Apps or want to show off what you're building, join us on the [Mattermost Apps channel in the Mattermost Community server](https://community.mattermost.com/core/channels/mattermost-apps)!
```
```suggestion
You now know how to create a Mattermost App in Python. If you have questions about building Apps or want to show off what you're building, join us on the {{<newtabref title="Mattermost Apps" href="https://community.mattermost.com/core/channels/mattermost-apps">}} channel in the Mattermost Community server!
```

neflyte commented on site/content/integrate/apps/quickstart/quick-start-serverless/_index.md:
```diff
@@ -90,4 +88,4 @@ Now that you've examined the App in HTTP mode, you can next look at deploying th
 
-You now know how to create a Mattermost App in Go. If you have questions about building Apps or want to show off what you're building, join us on the [Integrations & Apps channel in the Mattermost Community server](https://community.mattermost.com/core/channels/integrations)!
\ No newline at end of file
+You now know how to create a Mattermost App in Go. If you have questions about building Apps or want to show off what you're building, join us on the [Mattermost Apps channel in the Mattermost Community server](https://community.mattermost.com/core/channels/mattermost-apps)!
```
```suggestion
You now know how to create a Mattermost App in Go. If you have questions about building Apps or want to show off what you're building, join us on the {{<newtabref title="Mattermost Apps" href="https://community.mattermost.com/core/channels/mattermost-apps">}} channel in the Mattermost Community server!
```

neflyte commented on site/content/integrate/apps/quickstart/quick-start-ts/_index.md:
```diff
@@ -1,23 +1,23 @@
-weight: 20
+weight: 10
 aliases:
   - /integrate/apps/quick-start-js/
```
Please add the redirect for this page since its name has changed.

neflyte commented on site/content/integrate/apps/quickstart/quick-start-ts/_index.md:
```diff
@@ -1,23 +1,23 @@
+Before you can start with your App, you should first set up your environment by following the [developer setup guide]({{< ref "/integrate/apps/quickstart" >}}).
 
-You also need an [LTS version of NodeJS](https://nodejs.org/en/download/).
+You also need [NodeJS v16 (Gallium) LTS](https://nodejs.org/en/download/).
```
```suggestion
You also need {{<newtabref title="NodeJS v16 (Gallium) LTS" href="https://nodejs.org/en/download/">}}.
```

neflyte commented on site/content/integrate/apps/quickstart/quick-start-ts/_index.md:
```diff
@@ -86,4 +86,4 @@ docker compose down
 ## Conclusion
 
-You now know how to create a Mattermost App in TypeScript. If you have questions about building Apps or want to show off what you're building, join us on the [Integrations & Apps channel in the Mattermost Community server](https://community.mattermost.com/core/channels/integrations)!
+You now know how to create a Mattermost App in TypeScript. If you have questions about building Apps or want to show off what you're building, join us on the [Mattermost Apps channel in the Mattermost Community server](https://community.mattermost.com/core/channels/mattermost-apps)!
```
```suggestion
You now know how to create a Mattermost App in TypeScript. If you have questions about building Apps or want to show off what you're building, join us on the {{<newtabref title="Mattermost Apps" href="https://community.mattermost.com/core/channels/mattermost-apps">}} channel in the Mattermost Community server!
```

neflyte commented on site/content/integrate/apps/quickstart/_index.md:
```diff
@@ -1,18 +1,67 @@
 ---
-title: Quick start
-description: Guides for quickly getting started building apps
+title: Apps - Developer setup
```
I think the extra "Apps" looks redundant in the LHS. Would the reader not know they're in the Apps section?

azigler replied:
It's not redundant in the page title, which could benefit from the clarity. @mickmister makes a good point that if you open up a few tabs, it's hard to tell what is what in Chrome just by their names. I think this is something we would need to address across all of the docs, but I think it works okay here for now. 1/5

neflyte replied:
Agreed. :)

neflyte reviewed (changes requested):
Found another couple things; sorry for the multiple reviews.

cwarnermm commented:
@azigler - The dev docs site has seen a lot of changes recently. Given the age of this PR, would you be open to reviewing the changes proposed in this PR against content available in production to confirm whether we want to move ahead with this PR?

azigler commented:
@cwarnermm Thanks for checking about this one. From what I can tell, there are just tiny conflicts on two of the quickstart pages in this PR currently, and both can just be resolved in favor of this PR because it's a full revamp of those pages. Our current developer tutorials for apps are very spare, so I think getting this out soon (especially at the dawn of Hacktoberfest) is best, and then we can continue to iterate -- it's great to see we have a lot of ideas about improving these docs!

Despite the age, I don't think any of the content is expired, it seems to be working on my end. I see Alan has feedback about file permissions when running Docker, but that's a separate Docker issue outside of this guide that I can help address with him.

Let me know how you'd like to proceed!

cwarnermm commented:
Sounds great, @azigler! Would you be open to resolving the outstanding feedback and merge conflicts on this PR? Once resolved, we can merge this great update into production.

azigler commented on site/content/integrate/apps/quickstart/_index.md:
````diff
@@ -1,18 +1,67 @@
+Next, use the following command to bring your development environment online.
+
+```sh
+docker compose up -d
````
This is actually a separate issue that has to do with using the Docker image with certain file permissions. This can also happen if you try to have Docker use pre-existing folders (maybe folders that were generated during a previous start, and have different permissions). This has been reported by other users, and I have also experienced it.

For me, it's usually fixed by deleting the folders used by Dockers, making sure new folders are created and have the right perms (step #3 [here](https://docs.mattermost.com/install/install-docker.html)), and/or adding `user: 0:0` to my `docker-compose.yml`.

azigler commented:
/update-branch

azigler commented:
> Sounds great, @azigler! Would you be open to resolving the outstanding feedback and merge conflicts on this PR? Once resolved, we can merge this great update into production.

Do you hear that? The quiet stillness? That's the sound of 214 individually resolved conversations! This PR has been an incredible effort by everyone -- @neflyte @cwarnermm @mickmister these guides wouldn't exist without y'all's sagely perspectives! 🧙 

I resolved Alan's comments, incorporated the changes, and fixed the conflicts. I also merged master into this branch to update it. It failed to build at first, but then I fixed the link to @jordanafung's new contributor pages and now it's good!

I'd love to see this get merged to bolster Hacktoberfest app integrators and put resources in their hands ASAP. This is a night-and-day difference over what's currently there, and once it's live we'll continue to iterate because it's clear we all care very much about making these guides successful -- and I love it! ✨

@cwarnermm please let me know if there are any obstacles to merge at this time.
